### PR TITLE
Bug fix: Make FileStore respect the default artifact location

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?
 
Current FileStore behavior ignores the `artifact_uri` parameter of the class constructor when creating the default experiment (https://github.com/mlflow/mlflow/blob/aba3ab278ecf2d06c0d25703609db7410b63e00b/mlflow/store/file_store.py#L102). Additionally, the `FileStore.create_experiment()` method does not use the default artifact location specified in the `FileStore` constructor when an alternative artifact URI is not explicitly specified (`artifact_uri=None` is passed to `_create_experiment_with_id`: https://github.com/mlflow/mlflow/blob/aba3ab278ecf2d06c0d25703609db7410b63e00b/mlflow/store/file_store.py#L195-L196).

This PR fixes these bugs and introduces internal `artifact_uri` validation inside the `_create_experiment_with_id` method to validate that a `NoneType` is not passed when attempting to create an experiment.

Fixes #1332.
 
## How is this patch tested?
 
Extension of existing unit tests in the `test_rest_tracking` suite.

TODO: Consider adding additional tests to the `test_file_store` suite.
 
### Is this a user-facing change? 

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [X] Tracking
- [ ] Projects 
- [X] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [X] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by tomasatdatabricks — APPROVED** [review]
lgtm

File: mlflow/store/file_store.py [hunk 1]
```diff
@@ -190,10 +190,9 @@ def list_experiments(self, view_type=ViewType.ACTIVE_ONLY):
         return experiments
 
     def _create_experiment_with_id(self, name, experiment_id, artifact_uri):
+        assert artifact_uri, "experiment artifact URI must be a non-empty string"
```
Comment (mparkhe) [inline mlflow/store/file_store.py]: This `assert` is in production path.
If you intend this to be a check, can you if an `if not artifact_...` and raise an `MlflowException`

File: tests/tracking/test_rest_tracking.py [hunk 2]
```diff
@@ -363,7 +365,13 @@ def test_set_terminated_status(mlflow_client):
 
 def test_artifacts(mlflow_client):
     experiment_id = mlflow_client.create_experiment('Art In Fact')
```
Comment (mparkhe) [inline tests/tracking/test_rest_tracking.py]: haha!

File: mlflow/store/file_store.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/store/file_store.py]: Done! Thanks.

File: mlflow/store/file_store.py [hunk 3]
```diff
@@ -190,10 +190,11 @@ def list_experiments(self, view_type=ViewType.ACTIVE_ONLY):
         return experiments
 
     def _create_experiment_with_id(self, name, experiment_id, artifact_uri):
+        if not artifact_uri:
+            raise MlflowException(
+                "Experiment artifact URI must be a non-empty string, found: %s" % artifact_uri)
```
Comment (mparkhe) [inline mlflow/store/file_store.py]: `artifact_uri` is an empty-string or `None`

**Review by mparkhe — APPROVED** [review]
Small *nit*. LGTM otherwise...

File: mlflow/store/file_store.py (hunk 3 above)
Comment (dbczumar) [reply mlflow/store/file_store.py]: Can you clarify?

File: mlflow/store/file_store.py [hunk 4]
```diff
@@ -190,10 +190,11 @@ def list_experiments(self, view_type=ViewType.ACTIVE_ONLY):
...
         self._check_root_dir()
         meta_dir = mkdir(self.root_directory, str(experiment_id))
-        artifact_uri = artifact_uri or path_to_local_file_uri(
-            os.path.join(self.root_directory, str(experiment_id)))
```
Comment (aarondav) [inline mlflow/store/file_store.py]: Did we lose the logic which appends `experiment_id`? artifact_uri != root_artifact_uri. We should have a test to confirm this if this is a behavioral difference.

Comment (aarondav) [reply mlflow/store/file_store.py]: Btw, it seems that the R tests caught this behavioral change, in `test-tracking-experiments.R` :)

**Comment by dbczumar** [comment]
@aarondav @tomasatdatabricks Looks like the R tests were previously failing for 2 reasons:

1. As Aaron pointed out, the original PR removed critical logic for creating an artifact subdirectory with the experiment id.

2. During the `mlflow_list_experiments() works properly` test case, the `mlflow_list_experiments()` method in R communicates restfully to an MLflow server. The default arguments for initializing an MLflow server *specify* an artifact URI of `mlflow.store.DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH == ./mlruns` (https://github.com/mlflow/mlflow/blob/9a8ee6cde189206d65be76b0b62e5fe46a1422aa/mlflow/cli.py#L232). As a result, the `FileStore` is instantiated with a default artifact root URI of `./mlruns`, resulting in an experiment artifact location of `./mlruns/0`, which resolves to `"/home/travis/build/mlflow/mlflow/mlflow/R/mlflow/tests/testthat/mlruns/0"` in absolute form, instead of the test case's assumed URI of `file:///home/travis/build/mlflow/mlflow/mlflow/R/mlflow/tests/testthat/ml/runs/0`. Supporting log output obtained by augmenting the R tests with `expect_equal(experiments_list$artifact_location[[1], default_artifact_loc)`:
```
── 1. Failure: mlflow_list_experiments() works properly (@test-tracking-experime
experiments_list$artifact_location[[1]] not equal to `default_artifact_loc`.
1/1 mismatches
x[1]: "/home/travis/build/mlflow/mlflow/mlflow/R/mlflow/tests/testthat/mlruns/0"
y[1]: "file:///home/travis/build/mlflow/mlflow/mlflow/R/mlflow/tests/testthat/ml
y[1]: runs/0"
```

This test case passed previously because the FileStore's artifact root was erroneously ignored when creating the default experiment. A file URI based on the root directory was used instead.

In the future, we should consider whether the default MLflow server URI should have the `file://` scheme. For now, I've corrected the test to check for an artifact location of the form `default_artifact_loc <- file.path(getwd(), "mlruns", "0", fsep = "/")`, instead of `default_artifact_loc <- file.path("file:/", getwd(), "mlruns", "0", fsep = "/")`.

File: tests/tracking/test_rest_tracking.py [hunk 5]
```diff
@@ -364,7 +366,14 @@ def test_set_terminated_status(mlflow_client):
...
+    experiment_info = mlflow_client.get_experiment(experiment_id)
+    assert experiment_info.artifact_location.startswith(
+        path_to_local_file_uri(SUITE_ARTIFACT_ROOT_DIR))
+    artifact_path = urllib.parse.urlparse(experiment_info.artifact_location).path
+    assert posixpath.split(artifact_path)[-1] == experiment_id
```
Comment (dbczumar) [inline tests/tracking/test_rest_tracking.py]: @aarondav I added this test to verify that the artifact path ends with the experiment ID

File: mlflow/utils/file_utils.py [hunk 6]
```diff
@@ -355,7 +355,7 @@ def path_to_local_file_uri(path):
     Convert local filesystem path to local file uri.
     """
     path = pathname2url(path)
-    if path == posixpath.abspath(path):
+    if posixpath.isabs(path):
```
Comment (dbczumar) [inline mlflow/utils/file_utils.py]: Fixes an edge case I ran across while debugging where a path is not recognized as absolute if it ends in a slash.

**Comment by aarondav** [comment]
Well... now the Windows build is broken :(

**Pull request merged**